---
  html_document:
    toc: true
    highlight: zenburn
    theme: united
---
The code for this report is on github [here](https://github.com/hbc/rubin-cell-fate-conversion).

```{r setup, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
               cache=TRUE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
               message=FALSE, prompt=TRUE, comment='', fig.cap='')
```
```{r load-libaries}
library(useful)
library(readr)
library(dplyr)
library(tidyr)
library(ggplot2)
library(stringr)
```

Read in the data for each plate into one big dataframe, make unique IDs for
each batch/well combination and use those IDs as the column names.

```{r read-plates, eval=!file.exists("plates.RData")}
setwd("~/rubin-cell-fate-conversion")

# Creating a function to organize the data
plate_reader = function(fn) {
  batch = strsplit(basename(fn), ".", fixed=TRUE)[[1]][1]
  data = read.table(fn, header=TRUE, row.names=1)
  colnames(data) = paste(batch, colnames(data), sep="_")
  data
}
plates_fns = sort(list.files("data", pattern="\\.dat", full.names=TRUE))
plates = do.call(cbind, lapply(plates_fns, plate_reader))
save(plates, file="plates.RData")
```

```{r load-plates}
load("plates.RData")
```

Create a dataframe of the metadata about each sample. This has an identifier
for a sample, Which well it came from, which batch and what it was treated with.

```{r massage-well-identity-data}
welldata_fn = "metadata/Compound Layout 384w1B.csv"
welldata = read_csv(welldata_fn) %>%
  gather(column, treatment, -row) %>%
  mutate(well=paste(row, column, sep="")) %>%
  dplyr::select(well, treatment)
# Add column called experiment to keep or elimiate sph samples. I only know how to do this in excel.
write.csv(welldata, "~/rubin-cell-fate-conversion/welldata_preaddexpinfo.csv")
welldata = read_csv("~/rubin-cell-fate-conversion/welldata_postaddexpinfo.csv")
identities = data.frame(str_split_fixed(colnames(plates), "_", 3))
colnames(identities) = c("batch", "drop", "well")
identities$id = colnames(plates)
welldata = identities %>% dplyr::select(batch, well) %>%
         left_join(welldata, by="well")
rownames(welldata) = colnames(plates)
welldata$sample = colnames(plates)
```
# Now I want to remove the sph samples based on experiment to create a new df and appropriate welldata.... Below is how I would do this for one plate... Remove sph samples from dataframe
```{segragate Sph samples}
Sphplates = plates[, c(1,23,24,25,47,48,49,71,72,73,95,96,97,119,120,121,143,144,145,167,168,169,191,192,193,194,216,217,218,240,241,242,264,265,266,288,289,290,312,313,314,336,337,338,360,361,362,384)]
Sphwelldata = welldata[c(1,23,24,25,47,48,49,71,72,73,95,96,97,119,120,121,143,144,145,167,168,169,191,192,193,194,216,217,218,240,241,242,264,265,266,288,289,290,312,313,314,336,337,338,360,361,362,384),]
```

Verify that the samples match up between the read counts and the metadata dataframe
and make sure there are no NA counts.

```{r verify-data}
dim(Sphwelldata)
dim(Sphplates)
table(rownames(Sphwelldata) %in% colnames(Sphplates))
corner(Sphplates)
corner(Sphwelldata)
table(complete.cases(Sphplates))
```

Looks like we are good to go. While we're at it we will load the positive
control data.

```{r load-positive-control-data}
positive_fn = "data/Feo_positive_controls.unq.refseq.umi.SC_and_Myoblast_Raw.txt"
positive = read.table(positive_fn, header=TRUE, row.names=rownames(plates))
positive$id = NULL
positive_samples = data.frame(str_split_fixed(colnames(positive), "_", 3))
identities = data.frame(str_split_fixed(colnames(plates), "_", 3))$X1
positive_welldata = data.frame(batch=positive_samples$X1,
                               treatment=positive_samples$X1,
                               well=positive_samples$X1)
```

Now we will calculate some summary statistics about each sample. First let's look at how many reads originate from each sample.

```{r plot-reads-per-sample}
Sphwelldata$genes_detected = colSums(Sphplates > 0)
Sphwelldata$genes_detected_zscore = ave(Sphwelldata$genes_detected, FUN=scale)
Sphwelldata$genes_detected_pval = 2 * pnorm(-abs(Sphwelldata$genes_detected_zscore))
Sphwelldata$genes_detected_padj = p.adjust(Sphwelldata$genes_detected_pval, method="BH")
outliers = subset(Sphwelldata, genes_detected_padj < 0.1)
```

Now we will look at the number of genes 

```{r plot-genes-detected}
ggplot(outliers, aes(treatment, genes_detected, fill=batch)) +
   geom_bar(stat='identity', position='dodge') +
   facet_wrap(~batch) +
   geom_hline(yintercept=ave(welldata$genes_detected), linetype="dashed") +
   ylab("genes with counts < 0") +
   xlab("") +
   theme_bw() +
   scale_x_discrete(breaks=NULL)
```

Grahp the number of low count samples per plate. We can see this in boxplot

```{r low-counts-plate}
ggplot(Sphwelldata, aes(batch, genes_detected)) + geom_boxplot() +
   ylab("genes with counts > 0") +
   xlab("") +
   theme_bw()
```

Now we can show the total counts per sample in a box plot format

```{r boxplot-counts-plate}
Sphwelldata$counts = colSums(Sphplates)
ggplot(Sphwelldata, aes(batch, counts)) + geom_boxplot() +
   ylab("total counts") +
   xlab("") +
   theme_bw()
```

Now show this in a bar plot of each sample per plate

```{r boxplot-counts-sample}
ggplot(Sphwelldata, aes(well, counts, fill=batch)) +
   geom_bar(stat='identity', position='dodge') +
  geom_text(aes(label=treatment,hjust=0, vjust=-0.1),position=position_dodge(width = 1)) +
  facet_wrap(~batch) +
   geom_hline(yintercept=ave(Sphwelldata$genes_detected), linetype="dashed") +
   ylab("Counts per sample") +
   xlab("treatment") +
   theme_bw() +
   scale_x_discrete(breaks=NULL)
write.csv(welldata, "~/rubin-cell-fate-conversion/welldata.csv")
```

And the number of failed wells, where we call a well failed if it
had < 10,000 total counts.
This is currently ommitted as some of these samples had quite low reads
#```{r failed-wells}
#welldata$failed = colSums(plates) < 10000
#welldata %>% group_by(batch) %>% summarise(nfailed = sum(failed))
```

Now estimate library complexity

```{r plot-complexity}
ggplot(Sphwelldata, aes(counts, genes_detected, color=batch)) +
   geom_smooth(fill=NA) +
   ylab("genes with counts > 0") +
   xlab("total counts") +
   theme_bw()
```

In this histogram of the genes detected, we can see there are a set of
samples with a low amount of genes detected.

# Render HTML file of .rmd - Currently not working because later packages aren't installed yet.
# library(rmarkdown)
# render("Screenseq1B_analysis.Rmd")
```{r genes-detected-histogram}
ggplot(Sphwelldata, aes(genes_detected)) + geom_histogram() +
   theme_bw() + xlab("genes with counts > 0")
```

Here we drop those samples, using a cutoff of 7,500 genes.
Once again due to low reads we will omit the following scripts

```{r drop-low-genes-detected}
#welldata = subset(welldata, genes_detected > 7500)
#plates = plates[, rownames(welldata)]
```

```{r biomart-conversions}
library(biomaRt)
mouse = useMart(biomart = "ENSEMBL_MART_ENSEMBL",
                dataset="mmusculus_gene_ensembl",
                host = "jul2015.archive.ensembl.org")
conversions = getBM(attributes=c("ensembl_gene_id", "mgi_symbol", "gene_biotype"),
                    mart=mouse)
```

Samples do not have very many counts in noise genes, so that is not an issue.
Noise genes are flagged by either being too small to be picked up reliably in a
standard RNA-seq analysis or are highly variable and prone to introducing
distortion such as rRNA.

```{r add-noise-counts}
biotypes = unique(conversions$gene_biotype)
noise_rna_biotypes = c("Mt_tRNA", "Mt_rRNA", "snoRNA", "snRNA", "misc_RNA",
                       "scaRNA", "rRNA", "sRNA")
noise_rna_genes = subset(conversions, gene_biotype %in% noise_rna_biotypes)$mgi_symbol
noise_rna = rownames(Sphplates)[rownames(Sphplates) %in% noise_rna_genes]
Sphwelldata$noise_counts = colSums(Sphplates[noise_rna,])
ggplot(Sphwelldata, aes(batch, noise_counts)) + geom_boxplot() +
   ylab("counts in noise genes") +
   xlab("") +
   theme_bw()
```

We'll drop the noise genes from consideration even though there aren't many
counts in them.

```{r drop-noise-genes}
Sphplates = Sphplates[!rownames(Sphplates) %in% noise_rna_genes,]
```

We'll also drop all genes with that don't have at least 100 counts total and
are not seen in at least 4 samples. This cuts down the number of genes we
are considering to ~13,500.

This may be too stringent... considering something like MyoD or Pax7 may only show up in the postive controls and only in a few of the samples that are going through reprogramming... if this is the case the tol
total number of reads may never reach 100...
Omitted for the time being
#```{r drop-low-genes}
#plates = plates[rowSums(plates > 0) > 4 & rowSums(plates) > 60,]
#```

This is what we're left with in terms of samples:

```{r filtering-finish}
knitr::kable(Sphwelldata %>% group_by(batch) %>% summarize(total=n()))
```

and we're left with `r nrow(plates)` genes to consider.

The DESeq2 PCA is a little bit limited, it doesn't return the higher order
principal components and we are going to do t-SNE anyway
later on, so we'll load these counts into Seurat which has a more flexible
PCA implementation and also has a t-SNE implementation. We'll first look
at this data to try to figure out what is up with TCL3.

```{r seurat-setup}
install.packages("devtools")
library(devtools)
install_url("https://github.com/satijalab/seurat/releases/download/v1.4.0/Seurat_1.4.0.14.tgz", binary = TRUE)
library(Seurat)


seurat.raw.data = data.frame(log(Sphplates + 1))
seurat.data = new('seurat', raw.data=seurat.raw.data)
seurat.data = Setup(seurat.data, project="rubin", min.cells=1, min.genes=1000,
                    is.expr=1, meta.data=Sphwelldata)
```

Here we look at what are the most variable genes across the samples. We can
see a lot of these are subunits of ribosomal proteins, these will be used
for the PCA.

```{r mean-var-plot}
seurat.data = MeanVarPlot(seurat.data, fxn.x = expMean, fxn.y = logVarDivMean, y.cutoff = 1, x.low.cutoff = 1)
```

We'll focus on component 1 since this is typically a good starting point.

```{r pca-plot}
seurat.data = PCA(seurat.data, do.print=TRUE)
rot = seurat.data@pca.rot %>%
  tibble::rownames_to_column(var="sample")  %>%
  left_join(welldata, by="sample")
ggplot(rot, aes(PC1, PC2, label=treatment, color=batch)) +
  geom_point(size=4, show.legend=FALSE) +
  geom_point() +geom_text(aes(label=treatment),hjust=0, vjust=0) +
  theme_bw()
VizPCA(seurat.data,1:2)
PCHeatmap(seurat.data, pc.use = 1, do.balanced = FALSE)
```

Let's try tSNE to see what materializes.
```{ tSNE-plot}
TSNEseurat = RunTSNE(seurat.data, dim_embed = 3, perplexity = 1, max_iter=600)
TSNEPlot(TSNEseurat, 1, 2, pt.size=2)

```


```